keps/sig-node/runtime-class-scheduling.md

```diff
@@ -0,0 +1,344 @@
+
+### Goals
+
+- Pods using a RuntimeClass that is not supported by all nodes in a cluster are
+  automatically scheduled to nodes that support that RuntimeClass.
+- RuntimeClass scheduling is compatible with other scheduling constraints. For
+  example, a pod with a node selector for GPUs and a Linux runtime should be
+  scheduled to a linux node with GPUs (an intersection).
+
+### Non-Goals
+
+- Replacing [SchedulingPolicy](#scheduling-policy) with RuntimeClasses.
+
+The following are **currently** out of scope, but _may_ be revisited at a later
+date.
+
+- Automatic topology discovery or node labeling
+- Automatically selecting a RuntimeClass for a pod based on node availability.
+- Defining official or revserved label or taint schemas or for RuntimeClasses.
+
+## Proposal
+
+A new optional `Topology` structure will be added to the RuntimeClass API. The
+topology includes both `NodeSelector` rules and `Tolerations` that are mixed in
+to a pod using that RuntimeClass. The mix-in happens in the mutating admission
+phase, and is performed by a new `RuntimeController` built-in admission
+plugin. The same admission controller is shared with the [Pod Overhead][]
+proposal.
+
+[Pod Overhead]: https://github.com/kubernetes/enhancements/pull/887
+
+### User Stories
+
+#### Windows
+
+The introduction of [Windows nodes][] presents an immediate use case for
+heterogeneous clusters, where some nodes are running Windows, and some
+linux. From the inherent differences in the operating systems, it is natural
+that each will support a different set of runtimes. For example, Windows nodes
+may support Hyper-V sandboxing, while linux nodes support Kata-containers. Even
+native container support varies on each, with runc for Linux and runhcs for
+Windows.
+
+- As a **cluster administrator** I want to enable different runtimes on Windows
+  and Linux nodes.
+- As a **developer** I want to select a Windows runtime without worrying about
+  scheduling constraints.
+- As a **developer** I want to ensure my Linux workloads are not accidentally
+  scheduled to windows nodes.
+
+[Windows nodes]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/20190103-windows-node-support.md
+
+#### Sandboxed Nodes
+
+Some users wish to keep sandbox workloads and native workloads separate. For
+example, a node running untrusted sandboxed workloads may have stricter
+requirements about which trusted services are run on that node.
+
+- As a **cluster administrator** I want to ensure that untrusted workloads are
+  not colocated with sensitive data.
+- As a **developer** I want run an untrusted service without worrying about
+  where the service is running.
+- As a **cluster administrator** I want to autoscale trusted and untrusted nodes
+  independently.
+
+## Design Details
+
+### API
+
+The RuntimeClass definition is augmented with an optional `Topology` struct:
+
+```go
+type Topology struct {
```

Comment:
Do we have any limit of expressiveness when relying on `nodeSelectorRequirement` ? If so, do we plan address it here or rely on the schedulingPolicy ?

Reply:
Yes, we're dropping the ability to union (OR) different sets of requirements. If the different sets use the same label key, then this can be achieved using the "In" operator. Otherwise, I think suggesting that nodes are relabeled or requiring scheduling policy is a reasonable fallback.